### ISSUE TITLE
[TCIM][S2-1] 实现基础 tensor、F16/Q8 primitives 与投影

### ISSUE THREAD

**Issue by kkongling** [fixedpointworks/ds4#12]
Part of #10
前置 #11 已完成。

后续 HC、attention、router 和 FFN 共用基础计算。本票完成 O1–O4 的 typed device primitives；公开 `ds4_gpu_*` 接线按相邻注释移交 Step3。

## 范围

- 具体函数、alias/stride/返回契约见 `ds4_tcim.c` 的 S2-1 / O1–O4 注释；覆盖 embedding/HC 初始化、加法、RMS/head norm、F16 投影和 Q8_0 raw-block dot/投影。
- 复用 Step1 tensor/view、内存与提交生命周期、typed HDPL wrapper 和原 CPU/reference。覆盖单/多行、pair 独立输出、实际 GGUF 矩阵 shape、完整输出列和边界。
- 采用 S2-0 原冻结输入/golden/容差；IQ2/Q2、cache/router/routed MoE 归 #13。
- 沿用 #11 明载且用户确认的存放约定：Step2 专用 harness、fixture、golden、数值门和运行报告放 ignored `misc/`，复用现有基础设施，不要求为本票迁移到 tracked tests。实现及必要 runtime 声明/catalog/映射注释已提交。
- `xh2rt_add_f32` 的数值/非法输入已验；`ds4_gpu_add_tensor` 仍为 Step3 binding 待办，其失败断言应保留到真实接线完成。

## 完成与验证

- [x] F16 gather 解码与 HC streams 复制精确；合法 token/view、词表首尾、非法输入及越界防护。
- [x] norm/add 实际宽度、eps、alias/stride、零近零、单多行、非 tile 尾部和 finite；typed add 另有真实数值及非法输入补充。
- [x] F16 单双投影真实 shape；Q8 raw scale/signed bytes/block-row offset/tail 和实际矩阵；正常投影读取完整输出列。
- [x] 原始 S2-1 实现使用真实 GGUF 和 CPU 原冻结门验收：359 required +59 supplemental =418 次比较；另有 synthetic176，synthetic 不替代真实。
- [x] 完整129280词表列的9种rows和8个activation输出独立保存文件验收；明确原runner exit未捕获，不虚报exit0。
- [x] `make tcim` / `make test-tcim`、Step1、add、fake HAL 的 ASan/UBSan、错误和资源回归；相邻注释明确 verified primitive 与 Step3 binding 的边界。
- [x] 按上述存放约定保留可独立运行的 harness、原冻结门、命令、输出/误差与源码/镜像身份。

实现：`7436c45` / `385646a`。原始证据在 `misc/tcim-s2-1/`；2026-09-18 只读收尾审计在 `misc/tcim-s2-close-20260918/issue12/`。当前 `a52a290` 有界真实 F16/Q8 回归结果见[收尾评论](https://github.com/fixedpointworks/ds4/issues/12#issuecomment-5732852955)，不能把旧 snapshot 的完整覆盖自动归为当前镜像。

后续 Q8 优化镜像的扩大回归已在 #16 承接，目前351/359 required +59 supplemental，8个多行full-logits仍待补。它不否定本票原始标量完整验收，也不代表 O13 HC/norm/argmax 组合完成。关闭本票不关闭 #14/#15/#16 或 Step3，也不宣称完整 TCIM decode 可用。

**Comment by kkongling** [comment]
本票按已确认的 O1–O4 typed primitive 范围收尾；正文同步落实 #11 的既有专用验证材料留 ignored `misc/` 约定，以及 Step3 public binding 的边界。

实现 `7436c45` / `385646a` 已完成真实 GGUF **359 required +59 supplemental =418 次比较**，另有 synthetic176。此次只读重算四份 scalar TSV、原冻结门、完整词表验证记录与源码身份：全部 PASS、finite、未放宽门限。418是比较次数，唯一case ID为406（12个pair KV输出重复required ID）；不是418个独立case。

完整词表验证覆盖 rows `1/19/31/32/33/127/128/129/255`、每行129280列，97,606,400个矩阵float；17个矩阵/activation输出长度和SHA逐项等于原frozen。原长测runner退出码未捕获，此处依据保存的完整输出、独立校验和空闲资源收尾验收，**不声称原runner exit0**。该几何检查属于O4，不代替O13 HC/norm/argmax组合。

当前 `a52a290eb4fdde7ba71d10107afcc17bda2ec7f4` 已在独立snapshot补验 **6组、52 required +59 supplemental =111/111次比较**：8种真实F16形状的rows1/19/33（24项），Q8 shared_gate/up的单行加速、rows33原路径、完整129280列full-logits rows1、24项实际activation codes/scales，以及原59项raw/tail、RMS极值和单/多行KV+gate pair补充。6组runner及外层run-all均exit0，原门与预期集合逐项核对，原frozen未修改。52个required输出的max/RMS均0，root另对全部保存输出逐个校验SHA，与原frozen **52/52逐字节相同**；59补充最大max_abs为`1.1920929e-5`、最大case RMS为`1.33472633e-6`，分别通过各自原门（不跨算子合并为统一门）。每组收尾group/load0、DDR free24448 MB。

本轮当前源码的 `make -j2 tcim`、原 `--inspect` 和完整raw GGUF可读性检查均exit0：86,720,111,488 B完整读取，用时224.787秒；未增加模型身份门。原Makefile确认7份HDPL镜像最新，本远端源码/镜像校验一致，测试前后未变化。当前project.hex SHA为`a3b9dc8b87f580b5e0ff01730cab27f2e096edbcfdc2686f6c5b2251971efa39`，本次candidate SHA为`cf051a67887c74c4f7ac2b809f3c8c023ee062e7b5adc02b1aa6cf6c7c4bae63`。

本轮入口（先构建新目录，root统一设备调度，结果目录必须不存在）：
```sh
bash misc/tcim-s2-close-20260918/issue12-run/build-remote.sh REPO NEW_BUILD
bash misc/tcim-s2-close-20260918/issue12-run/run-all.sh REPO NEW_BUILD NEW_RESULTS
```
具体已执行命令、来源、原输出和汇总保留于 `/work/ds4/misc/tcim-s2-close-20260918/issue12-run/`，独立原输出SHA核对见同级`root/issue12-independent-output-check.json`。

此前同一最终镜像的 `make test-tcim` 的CLI/link/runtime及6个Q8+6个Q2永久case已通过，本次校验其源码/hex身份相同而未重复执行；Q2真实expert准入另见 #27；不把这些结果写成S2-1全套重跑。CPU reference以及fill/add/gather/rms实现自原始验收以来未变。

专用 harness、frozen、命令、结果继续按用户约定留在 `misc/tcim-s2-1/`；本轮独立审计在 `misc/tcim-s2-close-20260918/issue12/`。typed add 已有数值/非法输入验收，public `ds4_gpu_add_tensor` 接线仍属Step3，原失败断言保持正确；未将primitive冒称为binding。

后续优化镜像的8个多行full-logits扩大回归仍由 [#16 的承接记录](https://github.com/fixedpointworks/ds4/issues/16#issuecomment-5732310992) 跟踪；351/359+59绑定`efd1f81` snapshot，未冒称当前`a52a290`全量通过。#14/#15组合、#16聚合与Step3接线继续各自推进，不阻塞本票基础primitive范围闭合。
